### DISCUSSION TITLE
Test not catching url changes made with liveview?

### DISCUSSION THREAD

**Discussion by zampa00** [keathley/wallaby#672] [Q&A] [answered]
Hello,
this may be a lack of understanding on my side, I have a test that should navigate multiple pages
```
session
    |> visit(Routes.chat_room_path(@endpoint, :index)) # bring the user to http://localhost:4002/
    |> click(Query.link("New chat room"))
    |> assert(current_url == "http://localhost:4002/new"
```
and the link is a liveview patch that doesn't change the view
```
<div>
  <%= live_patch "New chat room", to: Routes.chat_room_path(@socket, :new) %>
</div>
```
If I run the code in the browser, it works, but the test always fail.

I can't find any note on different behaviour with liveview, but I may just have missed it.
Is my change page approach wrong?

**Comment by mhanberg** [comment]
assert is an ExUnit function. 

The code has you is basically saying `assert %Session{}, (nil == "http://localhost:4002/new)`. current_url/0 is not a function, so that will evaluate to nil. 

What you want is ` assert current_url(session) == "http://localhost:4002/new"`. 

This is the error that I suspect you're having, if you're note, please provide the actual error.

**Comment by zampa00** [comment]
You're right, I provided a bad example. Here's the correct wrong test:
```
current_url = session
    |> visit(Routes.chat_room_path(@endpoint, :index)) # bring the user to http://localhost:4002/
    |> click(Query.link("New chat room"))  # should bring the user to http://localhost:4002/new via a live_patch
    |> current_url()
assert(current_url == "http://localhost:4002/new")
```
that fails with
```
Assertion with == failed
     code:  assert current_url == "http://localhost:4002/new"
     left:  "http://localhost:4002/"
     right: "http://localhost:4002/new"
```

I think it's also worth noting that I tried to log the socket from the handle_params/3 function in my liveview, and it looks like it's not invoked from the `click(Query.link("New chat room"))`, even if it _can_ find the element.
When I try my code in the browser, it works as expected, that's the reason it leads me to think there's something wrong in my approach somewhere

**Reply by zampa00** [reply]
Update: I tried to generate a generic CRUD with `mix phx.gen.live Catalog Product products name:string:unique`, it creates a batch of pages 
```
live "/products", ProductLive.Index, :index
live "/products/new", ProductLive.Index, :new
live "/products/:id/edit", ProductLive.Index, :edit

live "/products/:id", ProductLive.Show, :show
live "/products/:id/show/edit", ProductLive.Show, :edit
```
and the /products page has a link defined as:
```
<span><%= live_patch "New Product", to: Routes.product_index_path(@socket, :new) %></span>
```
If I write the same test as
```
test "users can create products", %{session: session} do
  current_url = session
    |> visit("/products")
    |> click(Query.link("New Product"))
    |> current_url()

  assert(current_url == "http://localhost:4002/products/new")
end
```
it fails again with
```
Assertion with == failed
     code:  assert current_url == "http://localhost:4002/products/new"
     left:  "http://localhost:4002/products"
     right: "http://localhost:4002/products/new"
     stacktrace:
```

**Reply by mhanberg** [reply]
What happens if you take a screenshot after clicking the link or if you run the test in headed mode (where you can see the browser $

**Reply by zampa00** [reply]
Looks like a speed problem: the application doesn't update the URL fast enough, and current_url/1 is executed too early. Putting a take_screenshot/1 slows the test down just enough to have it not fail.
```
    new_session =
      session
      |> visit("/products")
      |> click(Query.link("New Product"))
      |> take_screenshot()

    assert(current_url(new_session) == "http://localhost:4002/products/new")
```
this test is green, but if I remove take_screenshot/1 it fails most of the times (but not always).

This test is always green:
```
session
      |> visit("/products")
      |> click(Query.link("New Product"))
      |> assert_has(xpath("//*[@id=\"product-form\"]"))
```

Some of my problems derived by WSL (version 1), it runs wallaby but sometimes it's not working as expected, for instance the above test fails consistently on my Windows machine, but works on my Mac. Also take_screenshot/1 seems to be broken on WSL, it crash with
```
** (RuntimeError) Wallaby had an internal issue with HTTPoison:
     %HTTPoison.Error{id: nil, reason: :timeout}
     code: take_screenshot(new_session)
     stacktrace:
       (wallaby 0.29.1) lib/wallaby/httpclient.ex:50: Wallaby.HTTPClient.make_request/5
       (wallaby 0.29.1) lib/wallaby/webdriver_client.ex:403: Wallaby.WebdriverClient.take_screenshot/1
       (wallaby 0.29.1) lib/wallaby/driver/log_checker.ex:6: Wallaby.Driver.LogChecker.check_logs!/2
       (wallaby 0.29.1) lib/wallaby/browser.ex:237: Wallaby.Browser.take_screenshot/2
       test/wal_web/live/user_visit_test.exs:23: (test)
```
but it's not a big deal, after all WIndows is not supported.

**Reply by mhanberg** [reply]
You are better off making an assertion on something in the page rather than the asserting the url. If you check for something in the page it should avoid the race condition.

**Comment by zampa00** [comment, marked as answer]
Wrap up answer: Wallaby does not support Windows. It _does_ seem to work with WSL1, but things are broken and does not behave as expected. I have not tried with WSL2 but I'm expecting similar results.
